### ISSUE TITLE
Every authoring write deletes the EDD sheet from the workbook, and the next rebuild compiles untyped

### ISSUE THREAD

**Issue by PaulSnow** [DTRules/DTRules#1094]
The Excel refresh writes decision tables and nothing else, so any workbook that
carries an EDD sheet loses it on every authoring write. Excel is the system of
record, so this deletes the entity data dictionary from the record.

## Reproduction — a write that changes nothing

```
$ dtrules build --from-excel                      # establish the manifest
$ dtrules table get Compute_Eligibility > t.json
$ dtrules table put Compute_Eligibility < t.json  # identical JSON back
```

| | sheets | EDD sheet |
|---|---|---|
| `CHIP.xlsx` before | 8 | yes |
| `CHIP.xlsx` after | 7 | **gone** |

Same on CorporateTax: `states/NV_corp.xlsx` goes from 6 sheets with an EDD to 5
without.

## It is not only a lost sheet

The next `build --from-excel` reads a workbook with no EDD, so the state's field
types are unknown and the DSL compiles untyped. Fixed-point operators silently
become integer ones:

```
before:  result.nv_gross_revenue result.nv_commerce_tax_threshold f<
after:   result.nv_gross_revenue result.nv_commerce_tax_threshold <
```

Both fields are `type="double"`. Comparing a scaled value against an unscaled
one reads them as the same magnitude, and a wrong comparison decides which
column fires — a wrong answer, not a rounded one.

The #1029 guard does not catch it: that refuses when the symbol table is
*empty*, and here it is merely missing the entities that lived in the destroyed
sheet.

## Age

Pre-existing, not introduced by #1078. `RefreshExcelIn` previously called
`ExportDecisionTables`, which is also decision-tables-only; #1078 corrected
*which* tables each workbook receives and did not change the fact that no EDD
is written.

## Fix

The refresh has to write the workbook as it found it — decision tables plus the
EDD for the entities belonging to that workbook. `WorkbookExporter.ExportCombined`
already writes both halves; the authoring refresh takes the DT-only path.

Worth a guard alongside: a refresh that reduces a workbook's sheet count should
be refused, the same way #1088 refuses an export that cannot carry every table.
Losing a sheet is losing rules.

Found while quoting the string literals in CorporateTax's NV classification
table (#1012).

**Comment by PaulSnow** [comment]
Most of this is fixed in #1095. Recording what is not, because the remainder is
a different problem than the one filed.

**Fixed.** The owned-tables export now writes the workbook's own entities beside
its tables, EDD files load recursively so nested `states/` layouts are seen, and
the numeric-downgrade guard covers comparisons (`f<`, `f>`, `f<=`, `f>=`, `f==`,
`f!=`) — it had the arithmetic operators and `cvfp` but not these, which is why
the `f<` → `<` downgrade went unreported.

A no-op authoring write now leaves the sheet count unchanged on CHIP, ChipApp,
KidAid, Poker, SinusitisTherapy, StateTax and Cribbage, all still verifying.
Before, CHIP lost a sheet on any write.

**Not fixed: CorporateTax.** Its workbooks still lose their EDD sheet, and the
reason is structural rather than a missed case.

All 51 state EDDs declare entities named `result` and `apportionment`. Entities
are keyed by name, so the 51 files collapse to **19 entities**, and querying
which of them belong to `NV_corp.xlsx` returns **none** — the surviving `result`
carries whichever `xls_file` won the load:

```
entities loaded: 19
NV_corp.xlsx entities: 0
distinct workbooks claimed: 3
```

So "the entities owned by this workbook" has no answer for this project. Each
state EDD is not declaring its own entities; it is contributing state-specific
fields to shared ones. The committed `NV_corp.xlsx` does have an EDD sheet, so
whatever wrote it originally had a notion of per-state entities that the loaded
model no longer preserves.

That wants deciding rather than patching: either an entity records every
workbook that contributed to it, or a workbook's EDD sheet is derived from the
EDD *file* that sits beside its tables rather than from entity ownership. The
second sounds closer to how these projects are actually laid out — `states/`
holds `NV_corp_dt.xml` and `NV_corp_edd.xml` as a pair — but it is a change to
what a workbook means, not a bug fix.

**Comment by PaulSnow** [comment]
**The per-state entity rename is the keystone. Everything else is blocked on
it, including work that looks unrelated.**

Deleting CorporateTax's 60 dead EDD fields is now possible through the
authoring API (#1099 made the state EDDs reachable). Done on a copy:

```
deleted=60 errors=0
tables 172 -> 172, conditions 303 -> 303, actions 476 -> 476
EDD fields 1846 -> 1786   (exactly the 60)
```

No rule content touched. But `verify` then reports **22 `[build]` findings**:
the XML lost the fields and the state workbooks did not. `CO_corp.xlsx` still
has its EDD sheet carrying them.

Refreshing that sheet needs `entitiesOwnedBy("CO_corp.xlsx")` to return CO's
entities, and it returns none — because the entity is named `result`, all 29
state EDDs declare a `result`, and they collapse to one whose `xls_file` is
whichever loaded last. That is this issue's remaining half, and it is now
demonstrably blocking:

| work | blocked by |
|---|---|
| delete 60 dead fields | state workbook EDD sheet cannot be refreshed |
| move 76 statutory constants into their tables | same — they are declared in state EDDs |
| the master-result fold (#1097) | same — three entities vanish on rebuild |
| per-state mappings (#1098 groundwork) | needs per-state entities to map into |

So the order is: **rename first.** With `nv_result` / `nv_apportionment`,
ownership is unambiguous by construction, each state workbook holds exactly its
own entities, and the other three items become ordinary edits.

The deletion is reproducible whenever the rename lands: 60 `delete-field` ops
against 22 state EDD files, derived by taking every field the state EDDs
declare and keeping those matched nowhere in any DT, map, core EDD or test
data.

Not committed — landing it as-is would put 22 samples into drift and fail the
gate that #1090 just turned on.

**Comment by PaulSnow** [comment]
**Pilot done on Nevada. The approach works, and it has to be done for all 51
states at once — a partial rename does not hold.**

## What the pilot proved

Renaming NV's entities through the authoring API, entirely within the contract:

- 23 operations against `states/NV_corp_edd.xml` — 2 `add-entity`, 19
  `add-field`, 2 `delete-entity` — 0 errors. `apportionment`/`result` become
  `nv_apportionment` (3 fields) / `nv_result` (16), with the `nv_` prefix moved
  off the field and onto the entity.
- 69 references rewritten across NV's 5 tables via `table put`
  (`result.nv_commerce_tax_rate` → `nv_result.commerce_tax_rate`). No
  old-style reference left.

And the thing this was all for:

```
NV_corp.xlsx before an authoring write: 5 sheets, no EDD   (stripped, nothing claimed it)
NV_corp.xlsx after the rename:          6 sheets, EDD present
```

Ownership becomes unambiguous by construction, so the workbook keeps its EDD
sheet. That is the remaining half of this issue, fixed — for the state that was
renamed.

`nv_apportionment.state_tax_rate` is also now Nevada's own rate rather than one
of 37 competing defaults on a shared field, which is the #1012 NV problem
dissolving as a side effect.

## Why it must be all 51 at once

A rebuild reverted the whole thing: the entities came back as
`apportionment`/`result` and the 17 table references with them.

29 state EDDs still declare `result` and 43 still declare `apportionment`.
Those collapse to one entity whose `xls_file` is whichever loaded last — and
the surviving declaration can name *any* state's workbook:

```
xls_file="CA_corp.xlsx"  xls_file="CO_corp.xlsx"  xls_file="DC_corp.xlsx"  ...
```

So while even one state still declares the shared name, the collided entity can
attach to a renamed state's workbook, and the import writes the old names back
over it. Renaming states one at a time cannot converge.

Core keeps its own `result` and `apportionment` for the genuinely shared fields
(`state_code` the dispatch key, and the working values the rules assign). Once
no state declares them there is no collision, and core's declaration owns
`CorporateTax_core.xlsx` cleanly.

## The operation

51 states × (2 add-entity + N add-field + 2 delete-entity), roughly 1,200 EDD
patches, plus a `table put` per state table to rewrite references — then one
rebuild. Every step through the authoring API; nothing hand-edited. The NV
script generalises directly: read the state's EDD, strip the two-letter prefix
onto the entity name, rewrite `(result|apportionment).<st>_X` to
`<st>_result.X` / `<st>_apportionment.X` using the state's own EDD to decide
which entity owns each field.

Not started, deliberately: a half-renamed project is worse than an unrenamed
one, and this wants to run and be validated in one go.

**Comment by PaulSnow** [comment]
**Correcting my previous comment: the rename does not have to be done for all
51 states at once. Renaming one state works and holds.**

I claimed a partial rename "cannot converge", on the theory that the merged
`result` entity carries one `xls_file` and could attach to a renamed state's
workbook, so another state's declaration would overwrite it. Paul asked how
California undoes something about Nevada. It does not, and I could not
reproduce the mechanism I described.

Clean run: rename Nevada only, leave the other 50 alone.

| | |
|---|---|
| NV EDD after two rebuilds | `nv_apportionment`, `nv_result` — held |
| NV table references | 17 new-style, 0 old-style — held |
| `verify` | 8 findings, all `[external]`, **zero `[build]`** |
| CA, CO, TX | untouched, still `apportionment` / `result` |

The revert I reported came from an earlier run that had a failed first rewrite
attempt — a wrong field-ownership map that matched nothing — and mixed binaries
built either side of #1100/#1101. I saw the end state, inferred a mechanism,
and wrote it up without isolating it.

## What this changes

The migration is incremental. One state at a time, each independently
verifiable, no coordinated cutover:

1. rename that state's entities in its EDD (`add-entity`, `add-field` ×N,
   `delete-entity` — 23 operations for NV)
2. rewrite its tables' references (`table put`, 69 for NV)
3. rebuild and verify that state

Core keeps `result` and `apportionment` for the genuinely shared fields
throughout, and states that have not been converted keep using them, which is
why the two coexist without interfering.

One caveat worth carrying forward: after the rename NV's workbook has no EDD
sheet, because `add-entity` does not set `xls_file` and nothing then claims the
workbook. It causes no drift — verify is clean — but it means the state's
entity definitions live only in XML until that is addressed, which is the
remaining half of this issue and now affects one state at a time rather than
blocking everything.

**Comment by PaulSnow** [comment]
**Mapping surface built (#1104), conversion runs clean, and it stops on the
constants question — which it has now demonstrated rather than argued.**

## Where the conversion got to

With `dtrules map` available, the whole thing runs through the authoring API:

```
33 states converted, 18 with nothing to convert, 0 failed operations
1,077 table references rewritten
39 new state entities registered in the mapping, 0 failed
verify: 7 findings, all [external], zero [build]
```

The `not defined by any Entity on the Entity Stack` failure is gone —
`add-entity` writes both the cardinality declaration and the initialization
push, so `ca_result` resolves. Entities go 19 → 58, each state workbook carries
its own EDD sheet, and nothing is lost: 1553 (entity, field) pairs before and
after once the rename is normalised.

## What it then exposed

CorporateTax's execution tests fail — and not because the conversion broke
them.

```
apportionment.state_tax = 800, want 88400   (1,000,000 x 8.84%)
```

800 is California's minimum franchise tax, taken because the computed tax came
out 0. Its rate is 0 because:

- `CA_corp_edd.xml` declares `apportionment.state_tax_rate` with
  `default_value="0.0884"` — California's real rate, carried as an **EDD
  default**
- **no CA rule ever sets it**
- 25 other states declare the same shared field with their own defaults

So California's tax was computed from whichever state's default won the merge,
and before the conversion that happened to be California's. Deleting some
states' `apportionment` declarations changed the winner, and CA fell back to
core's `0.0`.

The test was passing on load-order luck. The conversion did not break
California; it removed the accident that was hiding the defect.

## What finishing requires

The eight shared fields 37 states override are not one kind of thing:

| field | what it is |
|---|---|
| `state_tax_rate`, `economic_nexus_threshold`, `apportionment_formula`, `transaction_threshold` | **statutory constants** — belong to the state, as its own field or a literal in its table |
| `state_code` | the **dispatch key**, an input; the orchestrator builds table names from it. Must stay shared |
| `state_additions`, `state_credits`, `state_subtractions` | **working values** the rules assign per run. Stay shared |

Only the first row moves. That is the constants step, and the conversion cannot
land without it, because it is what currently supplies California's rate by
accident.

Not committed — the repo is back to green.

**Comment by PaulSnow** [comment]
**Constants move works — California passes. Two scenarios remain, and the
reason is narrower than before.**

## This round

The converter now also moves the statutory constants a state was carrying as
overriding EDD defaults — `state_tax_rate`, `economic_nexus_threshold`,
`apportionment_formula`, `transaction_threshold` — into the state's own entity,
and rewrites that state's references. `state_code` stays shared because the
orchestrator dispatches on it, and `state_additions` / `state_credits` /
`state_subtractions` stay shared because rules assign them per run.

```
51 states converted, 0 failed, 1,110 references rewritten
73 entities registered in the mapping, 0 failed
107 mapping attributes re-pointed, 0 failed
verify: 7 findings, all [external], zero [build]
```

California now computes 88,400 correctly, from `ca_apportionment.state_tax_rate
= 0.0884` — its own field rather than one of 25 competing defaults. That was
the headline blocker and it is gone.

## What is left

`ME_graduated_brackets` and `MT_flat_rate` still read zero. Their rules and
mappings are correct:

```
rule:    set me_result.taxable_income = max(me_result.apportioned_income + ...)
mapping: <setattribute tag='me_apportioned_income' RAttribute='apportioned_income'
                       enclosure='me_result'>
```

The input never arrives. The scenario nests the state's fields inside the
`<result>` element:

```xml
<result>
  <me_apportioned_income>4000000.0</me_apportioned_income>
```

so the data loader binds them to the `result` instance regardless of what the
mapping's `enclosure` says. Re-nesting them under `<me_result>` does not help
either — that element is not bound to an entity instance, so the values are
dropped.

So the enclosing element, not just the tag, decides where data lands. That
matters for the design: external data is not fully insulated from the internal
entity shape, and routing a tag to a different entity needs whatever binds an
element to an instance — the `createentity` declarations look like the place,
and `dtrules map` does not expose them yet.

That is the remaining piece: one more op on the map surface, plus the scenario
data for the two states that feed state-specific inputs. The other ~13
scenarios pass.

Not committed; the repo is green.

**Comment by PaulSnow** [comment]
**Separate loads work. The remaining question is narrow: which load call the
second document should use.**

Paul's point is right — the nesting constraint only exists if you insist on one
document. Loading a state's data separately removes it, and #1098 already made
the mapping an argument to loading rather than session state, so the mechanism
is there.

Tried it: split the ME scenario into a federal document and
`TestCase_ME_graduated.me.xml` holding `<me_result>` at its root, then loaded
both.

**It works.** With `LoadDataAndPushSingletons` for the second document,
`me_result.tier1_tax` and `tier2_tax` compute correctly — 12,250 and 55,510,
exactly the 1120ME bracket figures. The data reaches the entity through the
state's own document, which is the thing that could not be done before.

What is not right yet is the tail: `tier3_tax`, `tier4_tax` and
`tax_liability` come out zero, consistent with `taxable_income` landing at
1,050,000 instead of 4,000,000 — a bracket boundary, so part of the input is
arriving and part is not. Switching the second load to `LoadData` (not pushing
singletons again, since they are already on the stack) moves the boundary the
other way: tier1 passes and tier2 no longer does.

So both variants load *some* of the document. The question is how a second
document should bind to singletons already on the stack —
`LoadDataAndPushSingletons` builds a second set of instances, and `LoadData`
alone appears not to bind to the existing ones. `LoadDataAndPush(r, pushOrder)`
is the third option and takes an explicit order, which may be exactly the
control needed.

That is one focused question against `pkg/dtrules/mapping`, not a design
problem. Everything else is done:

```
51 states converted, 0 failed, 1,110 references rewritten
73 entities registered in the mapping, 107 attributes re-pointed, 0 failed
verify: 7 findings, all [external], zero [build]
California computes 88,400 from its own ca_apportionment.state_tax_rate
~13 of 15 scenarios pass
```

Not committed; repo is green.

**Comment by PaulSnow** [comment]
Both halves are merged.

**#1106 — the engine defect.** A workbook's tables were compiled against a symbol table holding only that workbook's own EDD; the project-wide map was seeded once and then written over by the first workbook carrying an EDD sheet. Unknown types fall back to integer, so any table reading an entity declared in another workbook compiled money as integers — `apportionment.state_credits f-` became `-`, `cvd` became `cvi`. The replacement also leaked forward, because one importer is reused for every workbook in the build loop with no save/restore.

Nothing reported it. The downgrade detector covered fixed-point losses and the `f<` comparisons added when NV hit this class, but not double arithmetic or `cvd` → `cvi`. Those entries are in now.

TaxReturn was regenerated: 219 assignments that truncated to integer now store as double, tests unchanged.

**#1107 — the sample.** Every state owns `xx_apportionment` and `xx_result`; the shared fields stay on `apportionment` in the core EDD. Along the way:

- The shared `apportionment` entity had drifted into **WY's** state EDD, owned by WY's workbook. Deleting Wyoming would have deleted the entity 50 other states read.
- **VT** read the shared 3.5/7.5/8.3% tier rates while its own entity already carried the correct 6.0/7.0/8.5% (32 V.S.A. § 5832).
- **SC** read the shared license fee, now `sc_apportionment.license_fee_minimum` = $15 (SC Code 12-20-50).

`dtrules map add-entity` now writes the tag request too. Without it an entity works but no document can be *rooted* at it — an unmapped element is unknown markup, so the loader skips it and every child tag inside it, silently. That is why the per-state data documents loaded as all zeroes while the rules ran perfectly: the child tags were mapped, and never seen, because their wrapper was not.

CA computes $88,400, ME's four-bracket arithmetic executes end to end for the first time, and CorporateTax verifies with zero `[build]` findings. The 7 remaining are the NV bare-word operators tracked in #1012.

**Comment by PaulSnow** [comment]
All three parts are merged:

- **#1105/#1078 line** — the Excel refresh writes the workbook's own entities alongside its tables, so an authoring write no longer deletes the EDD sheet. Confirmed by a no-op `table get | table put` on all 11 samples: the sheet survives and verify stays clean.
- **#1106** — the compile-untyped half. The symbol table was per-workbook, so any table reading an entity declared elsewhere compiled money as integers. It now layers the workbook's EDD over the project-wide map. TaxReturn regenerated: 219 truncating assignments became doubles.
- **#1108** — the guard you asked for alongside: an export that would leave a workbook without its EDD sheet is refused.

The guard is narrower than "a refresh that reduces a workbook's sheet count", because `table delete` legitimately removes one. Widening it to "no entity claims this workbook" needs #1109 first — TaxReturn's 50 state EDDs each declare the shared `result` entity naming their own workbook, and only one survives the merge.
